Source/WebCore/html/FormAssociatedCustomElement.cpp

```diff
@@ -0,0 +1,279 @@
+}
+
+void FormAssociatedCustomElement::disabledStateChanged()
+{
+    ASSERT(m_element->isDefinedCustomElement());
+    ValidatedFormListedElement::disabledStateChanged();
+    CustomElementReactionQueue::enqueueFormDisabledCallbackIfNeeded(*m_element, isDisabled());
+}
+
+void FormAssociatedCustomElement::didChangeForm()
+{
+    ASSERT(m_element->isDefinedCustomElement());
+    ValidatedFormListedElement::didChangeForm();
+    if (!m_formWillBeDestroyed)
+        CustomElementReactionQueue::enqueueFormAssociatedCallbackIfNeeded(*m_element, form());
+}
+
+void FormAssociatedCustomElement::willUpgrade()
+{
+    setDataListAncestorState(TriState::False);
+}
+
+void FormAssociatedCustomElement::didUpgrade()
+{
+    ASSERT(!form());
+
+    HTMLElement& element = asHTMLElement();
+
+    parseFormAttribute(element.attributeWithoutSynchronization(formAttr));
+    parseDisabledAttribute(element.attributeWithoutSynchronization(disabledAttr));
+    parseReadonlyAttribute(element.attributeWithoutSynchronization(readonlyAttr));
+
+    setDataListAncestorState(TriState::Indeterminate);
+    updateWillValidateAndValidity();
+    syncWithFieldsetAncestors(element.parentNode());
+    invalidateElementsCollectionCachesInAncestors();
+    restoreFormControlStateIfNecessary();
+}
+
+void FormAssociatedCustomElement::finishParsingChildren()
+{
+    if (!FormController::ownerForm(*this))
+        restoreFormControlStateIfNecessary();
+}
+
+void FormAssociatedCustomElement::invalidateElementsCollectionCachesInAncestors()
+{
+    auto invalidateElementsCache = [](HTMLElement& element) {
+        if (auto* nodeLists = element.nodeLists())
+            nodeLists->invalidateCachesForAttribute(HTMLNames::formAttr);
+    };
+
+    if (RefPtr<HTMLFormElement> form = this->form())
+        invalidateElementsCache(*form);
+
+    for (auto& ancestor : lineageOfType<HTMLFieldSetElement>(*m_element))
+        invalidateElementsCache(ancestor);
+}
+
+const AtomString& FormAssociatedCustomElement::formControlType() const
+{
+    return asHTMLElement().localName();
+}
+
+bool FormAssociatedCustomElement::shouldSaveAndRestoreFormControlState() const
+{
+    const auto& element = asHTMLElement();
+    ASSERT(element.reactionQueue());
+    return element.reactionQueue()->hasFormStateRestoreCallback();
+}
+
+FormControlState FormAssociatedCustomElement::saveFormControlState() const
+{
+    ASSERT(m_element->isDefinedCustomElement());
+
+    FormControlState savedState;
+
+    if (m_state.has_value()) {
+        // FIXME: Support File when saving / restoring state.
+        // https://bugs.webkit.org/show_bug.cgi?id=249895
+        bool didLogMessage = false;
+        auto logUnsupportedFileWarning = [&](RefPtr<File>) {
+            auto& document = asHTMLElement().document();
+            if (document.frame() && !didLogMessage) {
+                document.addConsoleMessage(MessageSource::JS, MessageLevel::Warning, "File isn't currently supported when saving / restoring state."_s);
```

Comment:
I'm not certain logging this to console is useful.
There isn't really anything developer can do in response to this error message.

Reply:
My intention was to help a developer to debug "why Safari doesn't call my callback while other browsers do" and, if they so need to save / restore `File`, they could encode / decode it to `string` manually.